### ISSUE TITLE
Linux Desktop Integration

### ISSUE THREAD

**Issue by hmaarrfk** [conda-forge/qt-main-feedstock#120]
### Comment:

This is a tracking issue for the many attempts at integrating qt "seamlessly" into the linux desktop.

* [ ] GTK3 Integration
     * One of 3 attempts will likely get merge (soon?).
     * CDT-first - https://github.com/conda-forge/qt-main-feedstock/pull/105
     * Subpackage for gtk3 plugin - https://github.com/conda-forge/qt-main-feedstock/pull/98
     * plugin shipped, but not GTK3 dependencies - https://github.com/conda-forge/qt-main-feedstock/pull/118
* [ ] Wayland integration
     * Xref: https://github.com/conda-forge/qt-main-feedstock/issues/116
     * Attempt at build: https://github.com/conda-forge/qt-main-feedstock/pull/119
* [ ] Gnome Plugin
    * This seems to be necessary to get nice looking window decorations on gnome.
    * https://www.reddit.com/r/gnome/comments/w27f63/why_window_decoration_waylandqt_looks_different/
    * https://github.com/FedoraQt/QGnomePlatform
* [ ] xkeyboard-config: https://github.com/conda-forge/qt-main-feedstock/issues/117

PS. I'm not familiar with desktop environments other than GTK, so I apologize if I missed anything.

**Comment by hmaarrfk** [comment]
@ccordoba12 sorry for having this conversation in a few different places but, I do think it is a little more than just "gtk3" integration.

In https://github.com/conda-forge/qt-main-feedstock/pull/98#issuecomment-1400663240 you raised the concern that users would have to "opt in" for this kind of integration.

I do know that I don't want to increase the packaging complexity of qt. To me this means really trimming down each qt-recipe to the bare minimum. There are a few bugs that I keep hitting with multiple output recipes, so I would really like to avoid that approach. I would much prefer using different feedstocks for each subcomponent of qt, especially things like theming.

I'm not sure how we would do this (reliably), without some kind of opt-in.

We need to detect:

- X11 vs Wayland
- GTK vs KDE vs Other

There seems to be a few things the users would be interested in integrating, but conda (to my knowledge) doesn't have great visibility into those system decisions.

I guess my question is:

- Is the automatic enabling of this desktop integration important for you at this stage?

I would prefer having well organized packages with minimal dependencies and waiting a little longer to get "better system detection". My main concerns with adding more "features" to a "core" package are expressed in https://github.com/conda-forge/conda-forge.github.io/issues/1894

**Comment by ccordoba12** [comment]
> Is the automatic enabling of this desktop integration important for you at this stage?

Well, I don't want to block your efforts on this direction (which I find really valuable), so if you think the automatic approach is harder to maintain, then I'm fine with it. My main concern was that non-experienced Linux users, and those that don't know that these extra features are available as Conda-forge packages, would never use them.

> I would prefer having well organized packages with minimal dependencies and waiting a little longer to get "better system detection".

Ok, no problem. Perhaps conda could recognize packages such as `__kde` or `__gnome` in the future (similar to the current `__win` or `__unix` ones) to automatically add these extra Qt packages.

**Comment by h-vetinari** [comment]
> There are a few bugs that I keep hitting with multiple output recipes, so I would really like to avoid that approach.

I maintain several multi-output recipes without much troubles. Which (kind of) bugs do you have in mind?

Just a sidenote though; since you're doing almost all of the work, if you prefer separate feedstocks and it works, why not!

**Comment by hmaarrfk** [comment]
For those interested in GTK integration, @jschueller @ccordoba12 @Tobias-Fischer a review of https://github.com/conda-forge/staged-recipes/pull/22086 would be appreciated.

You can build the package locally quite quickly and test it out yourself. or you can copy it from https://anaconda.org/mark.harfouche/qt-gtk-platformtheme

**Comment by Vanuan** [comment]
I wonder how are conflicts resolved. For example, if Qt relies on a specific version of libfontconfig, while gtk3 which is provided with a host uses a different libfontconfig. Does qt use system libraries as well?

**Comment by hmaarrfk** [comment]
To be honest, i'm not sure how the symbols are solved. I am pretty set personally on pushing forward with: https://github.com/conda-forge/staged-recipes/pull/22086

since I think it is easier to test, and to "roll back" in case of any mistake.

You should be able to test it with 5.15.6 at least from https://anaconda.org/mark.harfouche/qt-gtk-platformtheme

and we can move forward to updating it on conda-forge when it is merged through.

qt from conda-forge currently uses system libraries for X and OpenGL, though it has been proposed in https://github.com/conda-forge/qt-main-feedstock/pull/115 and the related PRs to move away from system X11 alltogether.

**Comment by BrenBarn** [comment]
Would this also theoretically handle issues like conda-forge/qt-feedstock#154 and conda-forge/pyqt-feedstock#87 ?  (Not that I have the remotest idea how they would be solved :-) )

**Comment by hmaarrfk** [comment]
Yes. I think both those issues should be resolved by installing the first package on that list.

**Comment by BrenBarn** [comment]
Okay, thanks.  Would that only be for Gnome/GTK?  Is the postscript at the end of the list meant to suggest that some similar package would be required for KDE but we don't know yet what package that might be?

**Comment by hmaarrfk** [comment]
Correct. No computer i manage uses anything other than ubuntu with Gnome.

I know nothing about other desktop environments

**Comment by Vanuan** [comment]
@BrenBarn I would believe Plasma already uses QFileDialog. I think the problem there is that it looks different. The plugin corresponding to QGnomePlatform is called KDEPlatformTheme: https://github.com/KDE/plasma-integration

**Comment by hmaarrfk** [comment]
> Not sure why it can't be a separate plugin

I think it can be a separate plugin from the GTK plugin. Is there some indication that it cannot.

I kept the GTK plugin outside of the current qt-main package to avoid an explosion of dependencies.

I think it should be doable with the KDE plugin too.

**Comment by Vanuan** [comment]
No, I meant why would QGnomePlatform and KDEPlatformTheme developers decided to embed the xdg-desktop-portal plugin (included in qtbase).

I think I've got it. You can't run both platform theme plugins at the same time. So if you want both GNOME styling/fonts and desktop portals you need them to be in a single platform theme plugin as there can only be a single platform theme running.

**Comment by hmaarrfk** [comment]
To integrate with:

* Ubuntu + X11 + qt5
```
mamba install qt-main qt-gtk-platformtheme
```

* Ubuntu + Wayland + qt5
```
mamba install qt-main qt-wayland qgnomeplatform qt-gtk-platformtheme
```

* Ubuntu + X11 + qt6
```
mamba install qt6-main qt6-gtk-platformtheme
```

* Ubuntu + Wayland + qt6
```
mamba install qt6-main qt6-wayland qgnomeplatform6 qt6-gtk-platformtheme
```

**Comment by hmaarrfk** [comment]
Update: it seems that the qgnomeplatform is now no longer maintained or active.

To integrate with:

* Ubuntu + qt5
```
# X11
mamba install qt-main qt-gtk-platformtheme
# With wayland
mamba install qt-main qt-wayland qt-gtk-platformtheme
```

* Ubuntu + qt6
```
# X11 only
mamba install qt6-main qt6-gtk-platformtheme
# With wayland
mamba install qt6-main qt6-wayland qt6-gtk-platformtheme
```

**Comment by Hofer-Julian** [comment]
Hey @hmaarrfk!

We are currently working on adding menuinst support for `pixi global` and wondered why qt applications refuse to start on Wayland DEs. I noticed that `qt-wayland` has to be injected in my environment for it to start.

Have you considered adding `qt-wayland` as a run-dependency of `qt-main`?

**Comment by hmaarrfk** [comment]
> Have you considered adding qt-wayland as a run-dependency of qt-main?

I think it is important to consider the fact that the issue I opened was in 2023. Back in 2023, Wayland was not in as good shape as it is today, and Qt's support for it was also very buggy.

Basically, I think the "omnibus" package of `qt-main` is going out the door with qt6 at conda-forge.

We are taking an approach where we split off qt6 into smaller packages for licensing reasons as well as users that wish to have a slimmer install.

PySide6 still hasn't resolved an issue I flagged for them a while back https://bugreports.qt.io/browse/PYSIDE-2787 I am not sure if you can help there. It might be up your alley.

I'm happy to have `qt6-wayland` be added more automagically if a `__wayland` meta package is added to conda. without that, I am very hesitant to push it forward.

Are you daily driving Wayland today? I try every few months, and it still gives me challenges in my own software and other software.

**Comment by hmaarrfk** [comment]
> refuse to start on Wayland DEs

this is somewhat strange. do you specifically uninstall X11 shims in your DE?

**Comment by hmaarrfk** [comment]
> Could you specify what makes you hesitant?

1. Once a package is "added" to an environment it is very hard to "remove" it with `mamba update` https://github.com/conda-forge/conda-forge.github.io/issues/1894
2. Some apps, like mine, work better through Qt+XWyaland rather than Qt+Wayland..
3. I think Qt + Wayland was quite buggy. I can't always test it due to that. Part of it is definitely because Ubuntu is always a little behind.
4. This https://bugreports.qt.io/browse/PYSIDE-2787 affects me personally since I have an other 3D engine managing my 3D visual.
5. Lack of personal testing: The last time I tried to daily drive Wayland, my system "cache" just kept increasing. I couldn't attribute it to any process. By the end of the day, of my 64GB of RAM, I only had "10" available for "non cache" (according to htop). I couldn't tell if it was abug in Fedora 41 or Ubuntu 24.10 (tried both) (both using Gnome). Searching on forums but couldn't find a relevant bug report.

I think it was definitely too early in 2023 to push it as default. I think it might still be too early in 2025. Maybe in 2026????

> Yes. Wayland+XWayland makes fewer problems for me than X11 alone.

I think this might be because of your hard work on Gnome apps ^_^

> Electron apps.

These are somewhat more important in my workflow than what I would want.....

**Comment by hmaarrfk** [comment]
I also think that nobody else was testing this desktop integration stuff on linux as a daily driver. So with some additional help. I"m glad to reconsider my "probably not ready in 2025" statement.
